Rel v6r12: dirac-dms-mirror

petricm opened the pull request:
Added rsync like utility to recursively copy directory to and from SE, with the optional exact copy (delete files in target directory that are not in source directory).

atsareg commented:
This is a very interesting command indeed. I have a couple of quick comments to it. First, the name of the command is a bit vague, may be better dirac-dms-directory-sync or something more indicating the directory synchronization. Second, please use the FileCatalog object rather than constructing it by means of the FileCatalogFactory. Third, it will be nice to factor out the utilities into a separate client class that can be used in different contexts: this command, DIRAC API, COMDIRAC command eventually, etc.

petricm commented:
I have renamed it to dirac-dms-directory-sync and replaced the FileCatalogFactory with FileCatalog object. I have also added multiprocessing functionality, which is handy when you have a greater number of smaller files. 

I can over time factor out the utilities into a separate client class, but for that I need a bit more time.

atsareg commented on DataManagementSystem/scripts/dirac-dms-directory-sync.py:
```diff
@@ -0,0 +1,552 @@
+  else:
+    return S_OK( 'Successfully uploaded file to %s' % storage )
+
+def downloadRemoteFile(dm, lfn, destination):
```
Why do you need a separate method for downloading/uploading which is exactly the same as the DataManager methods ?

atsareg commented on DataManagementSystem/scripts/dirac-dms-directory-sync.py:
```diff
@@ -0,0 +1,552 @@
+    else:
+      log.notice(threadLine+ " Downloading " + filename + " -> [DONE]")
+
+def runInParallel(arguments, listOfLists, function):
```
Please, see the ThreadPool utility of DIRAC for multithreaded execution

atsareg commented on DataManagementSystem/scripts/dirac-dms-directory-sync.py:
```diff
@@ -0,0 +1,552 @@
+  for process in processes:
+    process.join()
+
+  for process in processes:
```
In DIRAC we have a standard return structure for any bulk operation: S_OK( { "Successful": successful, "Failed": failed } ), where successful and failed are dictionaries with file names as keys and values is either True( or some useful string ) for successful and Error message for the failed. Please, take habit to use this structure.

atsareg commented on DataManagementSystem/scripts/dirac-dms-directory-sync.py:
```diff
@@ -0,0 +1,552 @@
+if __name__ == "__main__":
+  returnValue = run( args , sync, parallel )
+  if not returnValue['OK']:
+    gLogger.fatal(returnValue['Message'])
```
no point in fatal here, you exit anyway straight afterwards

atsareg commented on DataManagementSystem/scripts/dirac-dms-directory-sync.py:
```diff
@@ -0,0 +1,552 @@
+
+  res = syncDestinations( upload, source_dir, dest_dir, storage, delete, nthreads )
+  if not res['OK']:
+    return S_ERROR(res['Message'])
```
It can be just :
return res

chaen replied:
I will go even further: it MUST be :-)

atsareg commented on DataManagementSystem/scripts/dirac-dms-directory-sync.py:
```diff
@@ -0,0 +1,552 @@
+
+  return S_OK('Upload finished successfully')
+
+def chunkList(alist, nchunks):
```
Please, see Core.Utilities.List.breakListIntoChunks() utility

atsareg commented on DataManagementSystem/scripts/dirac-dms-directory-sync.py:
```diff
@@ -0,0 +1,552 @@
+      else:
+        gLogger.notice("Deleting "+ filename + " -> [DONE]")
+
+    for directoryname in result['Value']['Delete']['Directories']:
```
DIRAC coding conventions include Camel convention for naming variables:
directoryname -> directoryName

atsareg commented on DataManagementSystem/scripts/dirac-dms-directory-sync.py:
```diff
@@ -0,0 +1,552 @@
+        gLogger.notice("Deleting "+ directoryname + " -> [DONE]")
+
+
+  for directoryname in result['Value']['Create']['Directories']:
```
When uploading local files, DIRAC will create necessary remote directories automatically

atsareg commented on DataManagementSystem/scripts/dirac-dms-directory-sync.py:
```diff
@@ -0,0 +1,552 @@
+      gLogger.notice("Creating " + directoryname + " -> [DONE]")
+
+  listOfFiles = result['Value']['Create']['Files']
+  #Chech that we do not have to many threads
```
Orthography

atsareg commented on DataManagementSystem/scripts/dirac-dms-directory-sync.py:
```diff
@@ -0,0 +1,552 @@
+  try:
+    os.remove(path)
+  except OSError as e:
+    return S_ERROR('Directory creation failed:' + e.strerror)
```
Copy paste error message

atsareg commented on DataManagementSystem/scripts/dirac-dms-directory-sync.py:
```diff
@@ -0,0 +1,552 @@
+  Remove local directory
+  """
+  try:
+    os.rmdir(path)
```
This will not remove non-empty directory

atsareg commented on DataManagementSystem/scripts/dirac-dms-directory-sync.py:
```diff
@@ -0,0 +1,552 @@
+  return S_OK("Successfully removed directory")
+
+
+def createRemoteDirectory(fc,newdir):
```
This is usually not needed in DIRAC as the necessary directories are created automatically

atsareg commented on DataManagementSystem/scripts/dirac-dms-directory-sync.py:
```diff
@@ -0,0 +1,552 @@
+  """
+  Remove file from the catalog
+  """
+  storageElements = gConfig.getValue( 'Resources/StorageElementGroups/SE_Cleaning_List', [] )
```
There should be a smarter way to remove physically directories. Collect all the SEs of file replicas that were contained in this directory and subdirectories and remove from those SEs

atsareg commented:
I will merge this PR now but please take into account the comments for the further work on this command
